// @spectrum-web-components/core/dist/components/progress-circle/ProgressCircle.base.js
import { property as l, query as n } from "lit/decorators.js";
import { SizedMixin as b, SpectrumElement as h } from "@spectrum-web-components/core/shared/base/index.js";
import { getLabelFromSlot as d } from "@spectrum-web-components/core/shared/get-label-from-slot.js";
import { PROGRESS_CIRCLE_VALID_SIZES as u } from "./ProgressCircle.types.js";
var c = Object.defineProperty, r = (a, e, s, m) => {
  for (var t = void 0, o = a.length - 1, p; o >= 0; o--)
    (p = a[o]) && (t = p(e, s, t) || t);
  return t && c(e, s, t), t;
};
class i extends b(h, {
  validSizes: u
}) {
  constructor() {
    super(...arguments), this.indeterminate = !1, this.label = "", this.progress = 0;
  }
  makeRotation(e) {
    return this.indeterminate ? void 0 : `transform: rotate(${e}deg);`;
  }
  handleSlotchange() {
    const e = d(this.label, this.slotEl);
    e && (this.label = e);
  }
  firstUpdated(e) {
    super.firstUpdated(e), this.hasAttribute("role") || this.setAttribute("role", "progressbar");
  }
  updated(e) {
    super.updated(e), !this.indeterminate && e.has("progress") ? this.setAttribute("aria-valuenow", "" + this.progress) : this.hasAttribute("aria-valuenow") && this.removeAttribute("aria-valuenow"), e.has("label") && (this.label.length ? this.setAttribute("aria-label", this.label) : e.get("label") === this.getAttribute("aria-label") && this.removeAttribute("aria-label"));
    const s = () => !!(this.label || this.getAttribute("aria-label") || this.getAttribute("aria-labelledby") || this.slotEl.assignedNodes().length);
    window.__swc.DEBUG && !s() && this.getAttribute("role") === "progressbar" && window.__swc?.warn(
      this,
      "<sp-progress-circle> elements need one of the following to be accessible:",
      "https://opensource.adobe.com/spectrum-web-components/components/progress-circle/#accessibility",
      {
        type: "accessibility",
        issues: [
          'value supplied to the "label" attribute, which will be displayed visually as part of the element, or',
          "text content supplied directly to the <sp-progress-circle> element, or",
          'value supplied to the "aria-label" attribute, which will only be provided to screen readers, or',
          'an element ID reference supplied to the "aria-labelledby" attribute, which will be provided by screen readers and will need to be managed manually by the parent application.'
        ]
      }
    );
  }
}
r([
  l({ type: String, reflect: !0, attribute: "static-color" })
], i.prototype, "staticColor");
r([
  l({ type: Boolean, reflect: !0 })
], i.prototype, "indeterminate");
r([
  l({ type: String })
], i.prototype, "label");
r([
  l({ type: Number })
], i.prototype, "progress");
r([
  n("slot")
], i.prototype, "slotEl");
export {
  i as ProgressCircleBase
};
