// @spectrum-web-components/core/dist/components/badge/Badge.base.js
import { property as a } from "lit/decorators.js";
import { SizedMixin as u, SpectrumElement as l } from "@spectrum-web-components/core/shared/base/index.js";
import { ObserveSlotPresence as f } from "@spectrum-web-components/core/shared/observe-slot-presence.js";
import { ObserveSlotText as h } from "@spectrum-web-components/core/shared/observe-slot-text.js";
import { FIXED_VALUES as m, BADGE_VARIANTS_SEMANTIC as d } from "./Badge.types.js";
var A = Object.defineProperty, _ = Object.getOwnPropertyDescriptor, c = (r, t, e, i) => {
  for (var s = i > 1 ? void 0 : i ? _(t, e) : t, o = r.length - 1, n; o >= 0; o--)
    (n = r[o]) && (s = (i ? n(t, e, s) : n(s)) || s);
  return i && s && A(t, e, s), s;
};
class p extends u(
  h(f(l, '[slot="icon"]'), ""),
  {
    noDefaultSize: !0
  }
) {
  constructor() {
    super(...arguments), this.variant = "informative";
  }
  static {
    this.FIXED_VALUES = m;
  }
  static {
    this.VARIANTS_SEMANTIC = d;
  }
  get fixed() {
    return this._fixed;
  }
  set fixed(t) {
    if (t === this.fixed)
      return;
    const e = this.fixed;
    this._fixed = t, t ? this.setAttribute("fixed", t) : this.removeAttribute("fixed"), this.requestUpdate("fixed", e);
  }
  // ──────────────────────
  //     IMPLEMENTATION
  // ──────────────────────
  /**
   * @internal Used for rendering gap when the badge has an icon.
   */
  get hasIcon() {
    return this.slotContentIsPresent;
  }
  update(t) {
    if (super.update(t), window.__swc?.DEBUG) {
      const e = this.constructor;
      e.VARIANTS.includes(this.variant) || window.__swc.warn(
        this,
        `<${this.localName}> element expect the "variant" attribute to be one of the following:`,
        "https://opensource.adobe.com/spectrum-web-components/components/badge/#variants",
        {
          issues: [...e.VARIANTS]
        }
      ), "outline" in this && this.outline === !0 && !e.VARIANTS_SEMANTIC.includes(this.variant) && window.__swc.warn(
        this,
        `<${this.localName}> element only supports the outline styling if the variant is a semantic color variant.`,
        "https://opensource.adobe.com/spectrum-web-components/components/badge/#variants",
        {
          issues: [...e.VARIANTS_SEMANTIC]
        }
      );
    }
  }
}
c([
  a({ type: String, reflect: !0 })
], p.prototype, "variant", 2);
c([
  a({ reflect: !0 })
], p.prototype, "fixed", 1);
export {
  p as BadgeBase
};
