// @spectrum-web-components/core/dist/components/status-light/StatusLight.base.js
import { property as n } from "lit/decorators.js";
import { SizedMixin as a, SpectrumElement as p } from "@spectrum-web-components/core/shared/base";
var u = Object.defineProperty, c = (o, s, e, l) => {
  for (var t = void 0, i = o.length - 1, r; i >= 0; i--)
    (r = o[i]) && (t = r(s, e, t) || t);
  return t && u(s, e, t), t;
};
class d extends a(p, {
  noDefaultSize: !0
}) {
  constructor() {
    super(...arguments), this.variant = "info";
  }
  // ──────────────────────
  //     IMPLEMENTATION
  // ──────────────────────
  updated(s) {
    if (super.updated(s), window.__swc?.DEBUG) {
      const e = this.constructor;
      e.VARIANTS.includes(this.variant) || window.__swc.warn(
        this,
        `<${this.localName}> element expects the "variant" attribute to be one of the following:`,
        "https://opensource.adobe.com/spectrum-web-components/components/status-light/#variants",
        {
          issues: [...e.VARIANTS]
        }
      ), this.hasAttribute("disabled") && !("disabled" in this) && window.__swc.warn(
        this,
        `<${this.localName}> element does not support the disabled state.`,
        "https://opensource.adobe.com/spectrum-web-components/components/status-light/#states",
        {
          issues: [
            "disabled is not a supported property in Spectrum 2"
          ]
        }
      );
    }
  }
}
c([
  n({ type: String, reflect: !0 })
], d.prototype, "variant");
export {
  d as StatusLightBase
};
