// @spectrum-web-components/core/dist/components/divider/Divider.base.js
import { property as s } from "lit/decorators.js";
import { SizedMixin as u, SpectrumElement as f } from "@spectrum-web-components/core/shared/base/index.js";
import { DIVIDER_VALID_SIZES as c, DIVIDER_STATIC_COLORS as d } from "./Divider.types.js";
var n = Object.defineProperty, p = (r, t, a, v) => {
  for (var e = void 0, i = r.length - 1, o; i >= 0; i--)
    (o = r[i]) && (e = o(t, a, e) || e);
  return e && n(t, a, e), e;
};
class l extends u(f, {
  validSizes: c,
  noDefaultSize: !0
}) {
  constructor() {
    super(...arguments), this.vertical = !1;
  }
  static {
    this.STATIC_COLORS = d;
  }
  firstUpdated(t) {
    super.firstUpdated(t), this.setAttribute("role", "separator");
  }
  updated(t) {
    super.updated(t), t.has("vertical") && (this.vertical ? this.setAttribute("aria-orientation", "vertical") : this.removeAttribute("aria-orientation"));
  }
}
p([
  s({ type: Boolean, reflect: !0 })
], l.prototype, "vertical");
p([
  s({ reflect: !0, attribute: "static-color" })
], l.prototype, "staticColor");
export {
  l as DividerBase
};
