// @spectrum-web-components/core/dist/shared/base/Base.js
import { LitElement as f } from "lit";
import { version as p } from "./version.js";
const l = /* @__PURE__ */ new Set(), _ = () => {
  const i = document.documentElement.dir === "rtl" ? document.documentElement.dir : "ltr";
  l.forEach((c) => {
    c.setAttribute("dir", i);
  });
}, E = new MutationObserver(_);
E.observe(document.documentElement, {
  attributes: !0,
  attributeFilter: ["dir"]
});
const b = (i) => typeof i.startManagingContentDirection < "u" || i.tagName === "SP-THEME";
function v(i) {
  class c extends i {
    /**
     * @private
     */
    get isLTR() {
      return this.dir === "ltr";
    }
    hasVisibleFocusInTree() {
      const n = ((a = document) => {
        let t = a.activeElement;
        for (; t?.shadowRoot && t.shadowRoot.activeElement; )
          t = t.shadowRoot.activeElement;
        const s = t ? [t] : [];
        for (; t; ) {
          const o = t.assignedSlot || t.parentElement || t.getRootNode()?.host;
          o && s.push(o), t = o;
        }
        return s;
      })(
        this.getRootNode()
      )[0];
      if (!n)
        return !1;
      try {
        return n.matches(":focus-visible") || n.matches(".focus-visible");
      } catch {
        return n.matches(".focus-visible");
      }
    }
    connectedCallback() {
      if (!this.hasAttribute("dir")) {
        let e = this.assignedSlot || this.parentNode;
        for (; e !== document.documentElement && !b(
          e
        ); )
          e = e.assignedSlot || // step into the shadow DOM of the parent of a slotted node
          e.parentNode || // DOM Element detected
          e.host;
        if (this.dir = e.dir === "rtl" ? e.dir : this.dir || "ltr", e === document.documentElement)
          l.add(this);
        else {
          const { localName: n } = e;
          n.search("-") > -1 && !customElements.get(n) ? customElements.whenDefined(n).then(() => {
            e.startManagingContentDirection(this);
          }) : e.startManagingContentDirection(
            this
          );
        }
        this._dirParent = e;
      }
      super.connectedCallback();
    }
    disconnectedCallback() {
      super.disconnectedCallback(), this._dirParent && (this._dirParent === document.documentElement ? l.delete(this) : this._dirParent.stopManagingContentDirection(
        this
      ), this.removeAttribute("dir"));
    }
  }
  return c;
}
class C extends v(f) {
  static {
    this.VERSION = p;
  }
}
if (process.env.NODE_ENV === "development") {
  const i = {
    default: !1,
    accessibility: !1,
    api: !1
  }, c = {
    default: !1,
    low: !1,
    medium: !1,
    high: !1,
    deprecation: !1
  };
  window.__swc = {
    ...window.__swc,
    DEBUG: !0,
    ignoreWarningLocalNames: {
      ...window.__swc?.ignoreWarningLocalNames || {}
    },
    ignoreWarningTypes: {
      ...i,
      ...window.__swc?.ignoreWarningTypes || {}
    },
    ignoreWarningLevels: {
      ...c,
      ...window.__swc?.ignoreWarningLevels || {}
    },
    issuedWarnings: /* @__PURE__ */ new Set(),
    warn: (r, e, n, { type: a = "api", level: t = "default", issues: s } = {}) => {
      const { localName: o = "base" } = r || {}, u = `${o}:${a}:${t}`;
      if (!window.__swc.verbose && window.__swc.issuedWarnings.has(u) || window.__swc.ignoreWarningLocalNames[o] || window.__swc.ignoreWarningTypes[a] || window.__swc.ignoreWarningLevels[t])
        return;
      window.__swc.issuedWarnings.add(u);
      let m = "";
      s && s.length && (s.unshift(""), m = s.join(`
    - `) + `
`);
      const w = t === "deprecation" ? "DEPRECATION NOTICE: " : "", h = r ? `
Inspect this issue in the follow element:` : "", g = (r ? `

` : `
`) + n + `
`, d = [];
      d.push(
        w + e + `
` + m + h
      ), r && d.push(r), d.push(g, {
        data: {
          localName: o,
          type: a,
          level: t
        }
      }), console.warn(...d);
    }
  }, window.__swc.warn(
    void 0,
    "Spectrum Web Components is in dev mode. Not recommended for production!",
    "https://opensource.adobe.com/spectrum-web-components/dev-mode/",
    { type: "default" }
  );
}
export {
  C as SpectrumElement,
  v as SpectrumMixin
};
