// @spectrum-web-components/core/dist/shared/base/sizedMixin.js
import { property as z } from "lit/decorators.js";
var h = Object.defineProperty, u = Object.getOwnPropertyDescriptor, _ = (r, e, i, n) => {
  for (var s = u(e, i), o = r.length - 1, t; o >= 0; o--)
    (t = r[o]) && (s = t(e, i, s) || s);
  return s && h(e, i, s), s;
};
const m = {
  xxs: "xxs",
  xs: "xs",
  s: "s",
  m: "m",
  l: "l",
  xl: "xl",
  xxl: "xxl"
};
function d(r, {
  validSizes: e = ["s", "m", "l", "xl"],
  noDefaultSize: i,
  defaultSize: n = "m"
} = {}) {
  class s extends r {
    constructor() {
      super(...arguments), this._size = n;
    }
    static {
      this.VALID_SIZES = e;
    }
    get size() {
      return this._size || n;
    }
    set size(t) {
      const l = i ? null : n, c = t && t.toLocaleLowerCase(), p = e.includes(c) ? c : l;
      if (p && this.setAttribute("size", p), this._size === p)
        return;
      const x = this._size;
      this._size = p, this.requestUpdate("size", x);
    }
    update(t) {
      !this.hasAttribute("size") && !i && this.setAttribute("size", this.size), super.update(t);
    }
  }
  return _([
    z({ type: String })
  ], s.prototype, "size"), s;
}
export {
  m as ElementSizes,
  d as SizedMixin
};
